crates/safe/src/safe.nr

```diff
@@ -0,0 +1,519 @@
+use poseidon::poseidon2_permutation;
+use sha256::sha256_var;
+
+/// SAFE (Sponge API for Field Elements)
+///
+/// This module provides a complete implementation of the SAFE API in Noir as defined in:
+/// "SAFE (Sponge API for Field Elements) - A Toolbox for ZK Hash Applications"
+/// see https://hackmd.io/bHgsH6mMStCVibM_wYvb2w#22-Sponge-state for more details.
+///
+/// SAFE provides a unified interface for cryptographic sponge functions that can be
+/// instantiated with various permutations to create hash functions, MACs, authenticated
+/// encryption schemes, and other cryptographic primitives for ZK proof systems.
+///
+/// This implementation follows the SAFE specification exactly, providing:
+/// - Complete API: START, ABSORB, SQUEEZE, FINISH operations.
+/// - Full security: Domain separation, tag computation, IO pattern validation.
+/// - Poseidon2 integration: Field-friendly permutation for ZK systems.
+/// - Specification compliance: All operations follow SAFE spec 2.4 exactly.
+
+/// Rate parameter for the sponge construction (number of field elements that can be absorbed per permutation call).
+global RATE: u32 = 3;
+
+/// Capacity parameter for the sponge construction (security parameter, typically 1-2 field elements).
+global CAPACITY: u32 = 1;
+
+/// Total state size (rate + capacity) in field elements.
+global STATE_SIZE: u32 = RATE + CAPACITY;
+
+/// IO Pattern encoding constants (from SAFE spec 2.3).
+///
+/// These constants are used for encoding operation types in the 32-bit word format:
+/// - MSB set to 1 for ABSORB operations
```

Comment:
@YounesTal1 Do you remember why we removed this approach in the first version of SAFE? I recall doing something very similar (where 0 meant absorb and 1 meant squeeze in that version), but I think we decided this kind of approach wouldn’t be needed in our circuits. (I guess our zk circuits will call this function correctly without allowing anyone to change it to attack) Even though this doesn't affect circuit size that much, it may be optimized if we are sure that we won't need these checks